### PR TITLE
GSoC - SBML L3V2 models support with some other bug fixes

### LINKED ISSUE #38: SBML TestSuite failing test semantic/01488/01488 `java.lang.ArithmeticException: arcsech(0) undefined`
Test fails due to: `java.lang.ArithmeticException: arcsech(0) undefined`
```
java.lang.ArithmeticException: arcsech(0) undefined

	at org.sbml.jsbml.util.Maths.arcsech(Maths.java:145)
	at org.simulator.sbml.astnode.ASTNodeInterpreter.arcsech(ASTNodeInterpreter.java:727)
	at org.simulator.sbml.astnode.ASTNodeValue.computeDoubleValue(ASTNodeValue.java:408)
	at org.simulator.sbml.astnode.ASTNodeValue.compileDouble(ASTNodeValue.java:284)
	at org.simulator.sbml.astnode.ASTNodeInterpreter.functionDouble(ASTNodeInterpreter.java:192)
	at org.simulator.sbml.astnode.FunctionValue.computeDoubleValue(FunctionValue.java:121)
	at org.simulator.sbml.astnode.ASTNodeValue.compileDouble(ASTNodeValue.java:284)
	at org.simulator.sbml.astnode.ASTNodeInterpreter.functionDouble(ASTNodeInterpreter.java:190)
	at org.simulator.sbml.astnode.FunctionValue.computeDoubleValue(FunctionValue.java:121)
	at org.simulator.sbml.astnode.ASTNodeValue.compileDouble(ASTNodeValue.java:284)
	at org.simulator.sbml.astnode.RuleValue.processAssignmentVariable(RuleValue.java:128)
	at org.simulator.sbml.astnode.AssignmentRuleValue.processRule(AssignmentRuleValue.java:102)
	at org.simulator.sbml.SBMLinterpreter.processRules(SBMLinterpreter.java:2511)
	at org.simulator.sbml.SBMLinterpreter.init(SBMLinterpreter.java:1475)
	at org.simulator.sbml.SBMLinterpreter.<init>(SBMLinterpreter.java:564)
	at org.simulator.sbml.SBMLTestSuiteTest.testModel(SBMLTestSuiteTest.java:144)
	at sun.reflect.GeneratedMethodAccessor3.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runners.Suite.runChild(Suite.java:127)
	at org.junit.runners.Suite.runChild(Suite.java:26)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runners.Suite.runChild(Suite.java:127)
	at org.junit.runners.Suite.runChild(Suite.java:26)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:160)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

### LINKED ISSUE #41: Simulation of models with reaction without reactants and products
As the new SBML Level 3 Version 2 models now allow for reactions without reactants and products, this support has to be added to the SBSCL simulator. Currently, the simulation fails with a warning by giving NaN values in results.

The result after simulating the test case [1300](https://github.com/sbmlteam/sbml-test-suite/blob/develop/cases/semantic/01300/01300-sbml-l3v2.xml) from SBML Test Suite:
```
 WARN (SBMLCoreParser.java:477) - The reaction 'J0' does neither contain any reactants nor products.
[Time] [0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0], [p1], [NaN], [NaN], [NaN], [NaN], [NaN], [NaN], [NaN], [NaN], [NaN], [NaN], [NaN]]
```
A similar problem exists for test case [1301](https://github.com/sbmlteam/sbml-test-suite/blob/develop/cases/semantic/01301/01301-sbml-l3v2.xml) and [1302](https://github.com/sbmlteam/sbml-test-suite/blob/develop/cases/semantic/01302/01302-sbml-l3v2.xml).

### LINKED ISSUE #42: Add functions in ASTNodeInterpreter class for calculating node values of SBML L3V2 models
The SBML Level 3 Version 2 models contain some new math functions that need to be added to the SBSCL simulator. Currently, just NaN values are returned for these functions.
These are functions that need to be added:
- `FUNCTION_QUOTIENT`
- `FUNCTION_REMAINDER`
- `IMPLIES`
- `MAX`
- `MIN`

Some reference SBML L3V2 test cases: [1272](https://github.com/sbmlteam/sbml-test-suite/blob/master/cases/semantic/01272/01272-sbml-l3v2.xml)(Quotient), [1273](https://github.com/sbmlteam/sbml-test-suite/blob/master/cases/semantic/01273/01273-sbml-l3v2.xml)(Remainder), [1274](https://github.com/sbmlteam/sbml-test-suite/blob/master/cases/semantic/01274/01274-sbml-l3v2.xml)(Implies), [1275](https://github.com/sbmlteam/sbml-test-suite/blob/master/cases/semantic/01275/01275-sbml-l3v2.xml)(Max), and [1276](https://github.com/sbmlteam/sbml-test-suite/blob/master/cases/semantic/01276/01276-sbml-l3v2.xml)(Min).

### LINKED ISSUE #43: Add support for rateOf csymbol
A new `rateOf` csymbol has been introduced to the SBML Level 3 Version 2. Support for `rateOf` needs to be added to SBSCL which can be done by creating a function calculating a rate of any particular variable that needs to be added in the ASTNodeInterpreter class of SBSCL. Currently the SBML L3V2 models containing `rateOf` symbol outputs `NaN` values as results.

### PR DESCRIPTION
**Pull request by hemilpanchiwala**
### Major changes made in SBSCL (from GSoC start):
- Before this PR, two PRs have been merged to SBSCL. A complete description of changes made is described in that PRs.
    - PR #40 
    - PR #37 
- Absolute as well as Relative tolerances are now fixed for the simulation of SBML models. This has solved the problems of failing test cases of the SBML Test Suite due to small deviations.
- Now, SBML models with fbc extensions are simulated using FluxBalanceAnalysis in SBMLTestSuiteTest and SBMLTestRunnerWrapper.
- Active objective function variable is added to FluxBalanceAnalysis. And now all the fbc models in the SBML Test Suite pass successfully.
- SBML L3V2 models issues solved:
    - SBML L3V2 models do not have a `fast` property which will be handled now.
    - SBML L3V2 models now can have `constraints`, `kinetic laws`, `events`, `species`, `rules` (assignment rules and rate rules), `initial assignments`, and any others without math. This has been handled now through this PR in SBSCL.
    - Logical operators are now added to compileDouble in ASTNodeValue.
    - New functions are added to the ASTNodeInterpreter class. (issue #42)
    - Reaction without reactants and products has been solved now. (issue #41)
    - `RELATIONAL_GT`, `RELATIONAL_GEQ`, `RELATIONAL_EQ`, `RELATIONAL_NEQ`, `RELATIONAL_LEQ`, and `RELATIONAL_LT` are now updated to have multiple children.
    - SBSCL now supports `rateOf` csymbol that is newly added in SBML L3V2.
    - New variables have been added to `RateRuleValue` and `RuleValue` classes.
    - A new array `changeRate` has been added to SBMLinterpreter class that keeps the value of the derivatives at the current time.
    - `constantHash` has also been added to SBMLinterpreter that has key as variables and values are boolean (true if constant else false).
    - After making this changes, about 67 SBML L3V2 tests fail excluding the comp tests from the SBML Test Suite. 
- Many other small issues have been resolved in this PR.

This PR closes #38, closes #41, closes #42, closes #43.
More information about changes made till now in SBSCL can be found at my blog posts [[Link](https://gsoc20hemil.blogspot.com/)].

### CODE REVIEW

File: src/main/java/org/simulator/sbml/SBMLinterpreter.java [hunk 1]
```diff
@@ -69,25 +70,7 @@
...
-import org.simulator.sbml.astnode.PlusValue;
-import org.simulator.sbml.astnode.PowerValue;
-import org.simulator.sbml.astnode.RateRuleValue;
-import org.simulator.sbml.astnode.ReactionValue;
-import org.simulator.sbml.astnode.RootFunctionValue;
-import org.simulator.sbml.astnode.SpeciesReferenceValue;
-import org.simulator.sbml.astnode.SpeciesValue;
-import org.simulator.sbml.astnode.StoichiometryValue;
-import org.simulator.sbml.astnode.TimesValue;
+import org.simulator.sbml.astnode.*;
```
Comment (matthiaskoenig) [inline src/main/java/org/simulator/sbml/SBMLinterpreter.java:73]: Don't do `*` imports. This is normally not very good style and clutters the namespace.

**Comment by hemilpanchiwala** [comment]
> You have a ton of formating changes of comments in this pull request which makes it very hard to review the actual changes. In the future do separate pull requests for formating changes (of the 98 changed files only ~5-6 were actually changed)

@matthiaskoenig, will take care of it from next time.

File: src/main/java/org/simulator/examples/SBMLTestSuiteRunnerWrapper.java [hunk 2]
```diff
@@ -52,7 +55,7 @@ public static void main(String[] args) throws IOException, XMLStreamException, M
         String currentCase = args[1];
         String outputDirPath = args[2];
         String level = args[3];
-        String  version = args[4];
+        String version = args[4];
```
[the remark this answers is not in the document]
Comment (hemilpanchiwala) [reply src/main/java/org/simulator/examples/SBMLTestSuiteRunnerWrapper.java]: @matthiaskoenig, I am storing in amountHash as SBMLinterpreter takes amountHash as a parameter of its constructor.

File: src/test/scripts/sbml_test_suite_runner_wrapper.sh [hunk 3]
```diff
@@ -9,4 +9,4 @@
 # lib directory
 DIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
 
-java -cp $DIR/../../../target/classes/:$DIR/../../../src/lib/sbml_test_runner_wrapper/* org.simulator.examples.SBMLTestSuiteRunnerWrapper $1 $2 $3 $4 $5
+java -cp $DIR/../../../target/classes/:$DIR/../../../src/lib/sbml_test_runner_wrapper/*:$DIR/../../../src/lib/maven/kisao/libkisao/1.0.3.1-rc/libkisao-1.0.3.1-rc.jar:$DIR/../../../src/lib/maven/scpsolver/GLPKSolverPack/4.35v2/GLPKSolverPack-4.35v2.jar:$DIR/../../../src/lib/maven/scpsolver/LPSOLVESolverPack/5.5.2.5/LPSOLVESolverPack-5.5.2.5.jar:$DIR/../../../src/lib/maven/scpsolver/SCPSolver/1.0v2/SCPSolver-1.0v2.jar org.simulator.examples.SBMLTestSuiteRunnerWrapper $1 $2 $3 $4 $5
```
[the remark this answers is not in the document]
Comment (hemilpanchiwala) [reply src/test/scripts/sbml_test_suite_runner_wrapper.sh:12]: @matthiaskoenig, These new libraries I stated were already present in the SBSCL directory and are used for simulating FBC models, so I added their path explicitly. 

For making a general import I will have to copy the same libraries to the folder `sbml_test_runner_wrapper`.

File: src/main/java/org/simulator/math/MaxDivergenceTolerance.java [hunk 4]
```diff
@@ -6,102 +6,95 @@
...
- *             <mi>j</mi>
- *           </mrow>
- *         </msub>
- *         <mo>|</mo>
- *       </mrow>
- *     </mrow>
- *   </mfrac>
- *   <mo>&#x2264;<!-- ≤ --></mo>
- *   <mn>1</mn>
+ * <mfrac>
```
[the remark this answers is not in the document]
Comment (hemilpanchiwala) [reply src/main/java/org/simulator/math/MaxDivergenceTolerance.java:13]: @matthiaskoenig, I have made a formatting change according to the JSBML templates. I will update it.

File: src/main/java/org/simulator/sbml/SBMLinterpreter.java (hunk 1 above)
Comment (hemilpanchiwala) [reply src/main/java/org/simulator/sbml/SBMLinterpreter.java]: Updated

File: src/main/java/org/simulator/math/MaxDivergenceTolerance.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (hemilpanchiwala) [reply src/main/java/org/simulator/math/MaxDivergenceTolerance.java:59]: Updated

File: src/main/java/org/simulator/examples/SBMLTestSuiteRunnerWrapper.java [hunk 5]
```diff
@@ -5,16 +5,17 @@
 import org.sbml.jsbml.SBMLDocument;
```
[the remark this answers is not in the document]
Comment (hemilpanchiwala) [reply src/main/java/org/simulator/examples/SBMLTestSuiteRunnerWrapper.java]: Updated

File: src/main/java/org/testsuite/SBMLTestSuiteRunnerWrapper.java [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+
+                SBMLinterpreter interpreter = new SBMLinterpreter(model, 0, 0, 1, amountHash);
+
+                // Compute the numerical solution of the problem
+                solution = solver.solve(interpreter, interpreter.getInitialValues(), timepoints);
+            } else {
+                CompSimulator compSimulator = new CompSimulator(sbmlfile);
+                double stepSize = (duration / steps);
+
+                solution = compSimulator.solve(stepSize, duration);
```
Comment (zakharc) [inline src/main/java/org/testsuite/SBMLTestSuiteRunnerWrapper.java:172]: If we will not provide any changes to the CompSimulator one needs to deal with exceptions here (we are throwing except. in the corresponding solve method). Please consider this.

File: src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java [hunk 7]
```diff
@@ -37,30 +39,33 @@
...
 
     @Before
-    public void setUp(){ }
+    public void setUp() {
+    }
 
     @After
-    public void tearDown(){ }
+    public void tearDown() {
+    }
```
Comment (zakharc) [inline src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java:50]: Do we need to override these methods if we do nothing? As far as I can remember, JUnitTestRunner does nothing if there is no @Before, @After, @SetUp, or @TearDown methods are specified.

File: src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java [hunk 8]
```diff
@@ -159,7 +165,73 @@ public void testModel() throws FileNotFoundException, IOException, XMLStreamExce
...
-                if (model.getExtension(CompConstants.shortLabel) == null){
+                if (model.getExtension(CompConstants.shortLabel) != null) {
+
+                    // initialize simulator
+                    CompSimulator compSimulator = null;
+                    boolean errorInCompSimulator = false;
+                    try {
+                        compSimulator = new CompSimulator(sbmlFile);
+                    } catch (Exception e) {
+                        errorInCompSimulator = true;
```
Comment (zakharc) [inline src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java:176]: This is not good. Why we would suppress the exception in the test? That is why we write the tests. If everything is working - we are satisfied, if not - we want to know the reason. Please feel free to throw an exception in the tests. JUnit will treat this as an error and you will be able to see the stack trace to solve the problem fast and not spend your time guessing why this problem occurs.

File: src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java [hunk 9]
```diff
@@ -159,7 +165,73 @@ public void testModel() throws FileNotFoundException, IOException, XMLStreamExce
...
+                    }
+                    Assert.assertNotNull(compSimulator);
+                    Assert.assertFalse(errorInCompSimulator);
+
+                    //solve
```
Comment (zakharc) [inline src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java:181]: I am not sure if this comment necessary here or explains something.

File: src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java [hunk 10]
```diff
@@ -159,7 +165,73 @@ public void testModel() throws FileNotFoundException, IOException, XMLStreamExce
...
+                    MultiTable solution = null;
+                    boolean errorInSolve = false;
+                    try {
+                        double stepSize = (duration / steps);
+                        solution = compSimulator.solve(stepSize, duration);
+                    } catch (Exception e) {
+                        errorInSolve = true;
```
Comment (zakharc) [inline src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java:188]: Please see related comment above.

File: src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java [hunk 11]
```diff
@@ -159,7 +165,73 @@ public void testModel() throws FileNotFoundException, IOException, XMLStreamExce
...
+                    //TODO: Add quality measure to check whether solution meets the correct results
+
+                } else if (model.getExtension(FBCConstants.shortLabel) != null) {
+
+                    FluxBalanceAnalysis solver = null;
+                    boolean errorInFBASimulator = false;
+                    try {
+                        solver = new FluxBalanceAnalysis(org.sbml.jsbml.SBMLReader.read(sbmlFile));
+                    } catch (Exception e) {
+                        errorInFBASimulator = true;
```
Comment (zakharc) [inline src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java:202]: Please see related comment above.

File: src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java [hunk 12]
```diff
@@ -159,7 +165,73 @@ public void testModel() throws FileNotFoundException, IOException, XMLStreamExce
...
+                    Assert.assertFalse(errorInSolve);
+
+                    //TODO: Add quality measure to check whether solution meets the correct results
+
+                } else if (model.getExtension(FBCConstants.shortLabel) != null) {
+
+                    FluxBalanceAnalysis solver = null;
+                    boolean errorInFBASimulator = false;
+                    try {
+                        solver = new FluxBalanceAnalysis(org.sbml.jsbml.SBMLReader.read(sbmlFile));
```
Comment (zakharc) [inline src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java:200]: Just curious: why we use the full canonical name here? Do we bind the corresponding library dynamically?

File: src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java [hunk 13]
```diff
@@ -159,7 +165,73 @@ public void testModel() throws FileNotFoundException, IOException, XMLStreamExce
...
+                        errorInFBASimulator = true;
+                    }
+                    Assert.assertNotNull(solver);
+                    Assert.assertFalse(errorInFBASimulator);
+
+                    boolean errorInSolve = false;
+                    try {
+                        solver.solve();
+                    } catch (Exception e) {
+                        errorInSolve = true;
```
Comment (zakharc) [inline src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java:211]: Please see related comment above.

File: src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java [hunk 14]
```diff
@@ -159,7 +165,73 @@ public void testModel() throws FileNotFoundException, IOException, XMLStreamExce
...
+                    String[] values = reader.readLine().trim().split(",");
+
+                    Map<String, Double> inputSolution = new HashMap<>();
+                    for (int i = 0; i < keys.length; i++) {
+                        inputSolution.put(keys[i], Double.valueOf(values[i]));
+                    }
+
+                    for (Map.Entry<String, Double> mapElement : inputSolution.entrySet()) {
+                        if (solution.containsKey(mapElement.getKey())) {
+                            Assert.assertEquals(mapElement.getValue(), solution.get(mapElement.getKey()), 0.0002);
```
Comment (zakharc) [inline src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java:230]: Whatever this magical number means (0.0002) - I would exclude it to a global var.

File: src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java [hunk 15]
```diff
@@ -159,7 +165,73 @@ public void testModel() throws FileNotFoundException, IOException, XMLStreamExce
...
+                    try {
+                        solver.solve();
+                    } catch (Exception e) {
+                        errorInSolve = true;
+                    }
+                    Assert.assertFalse(errorInSolve);
+
+                    Map<String, Double> solution = solver.getSolution();
+
+                    System.out.println(solution);
```
Comment (zakharc) [inline src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java:217]: Please see related comment above.

File: src/test/scripts/sbml_test_suite_runner_wrapper.sh [hunk 16]
```diff
@@ -9,4 +9,4 @@
 # lib directory
 DIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
 
-java -cp $DIR/../../../target/classes/:$DIR/../../../src/lib/sbml_test_runner_wrapper/* org.simulator.examples.SBMLTestSuiteRunnerWrapper $1 $2 $3 $4 $5
+java -cp $DIR/../../../target/classes/:$DIR/../../../src/lib/sbml_test_runner_wrapper/*:$DIR/../../../src/lib/maven/kisao/libkisao/1.0.3.1-rc/libkisao-1.0.3.1-rc.jar:$DIR/../../../src/lib/maven/scpsolver/GLPKSolverPack/4.35v2/GLPKSolverPack-4.35v2.jar:$DIR/../../../src/lib/maven/scpsolver/LPSOLVESolverPack/5.5.2.5/LPSOLVESolverPack-5.5.2.5.jar:$DIR/../../../src/lib/maven/scpsolver/SCPSolver/1.0v2/SCPSolver-1.0v2.jar org.testsuite.SBMLTestSuiteRunnerWrapper $1 $2 $3 $4 $5
```
Comment (zakharc) [inline src/test/scripts/sbml_test_suite_runner_wrapper.sh:12]: This looks scary :)

Can we deal with this in a more generic way? For example: exporting system global variables and running this command with them as arguments.

**Comment by hemilpanchiwala** [comment]
> @hemilpanchiwala: Some of the comments are not related to your code. Please address the comments related to your changes first. For the rest: I am OK if you will adjust the rest during the next phase.
> 
> Small comments for the future:
> 
> * Please always exclude the formatting changes to the separate PR. I've seen a lot of whitespace changes/etc. in this PR. This will only add work for the reviewer and is not interesting at all.
> * I would suggest looking into coding conventions and guidelines of Eclipse Foundation (if you didn't do this already): https://wiki.eclipse.org/Development_Conventions_and_Guidelines
> 
> Nice work.

@zakharc, Extremely sorry for adding the formatting in this PR. I will take care from next time. And, I have formatted the code according to the template provided by the JSBML community linked [here](https://github.com/sbmlteam/jsbml/tree/master/dev).

File: src/main/java/org/simulator/comp/CompSimulator.java [hunk 17]
```diff
@@ -16,78 +16,80 @@
...
+    CompFlatteningConverter compFlatteningConverter = new CompFlatteningConverter();
+    docFlat = compFlatteningConverter.flatten(doc);
+    assert docFlat != null;
+  }
+
+  public SBMLDocument getDoc() {
+    return doc;
+  }
+
+  public SBMLDocument getDocFlat() {
```
[the remark this answers is not in the document]
Comment (hemilpanchiwala) [reply src/main/java/org/simulator/comp/CompSimulator.java:56]: @zakharc, or maybe getFlattenedDoc?

[the remark this answers is not in the document]
Comment (zakharc) [reply src/main/java/org/simulator/comp/CompSimulator.java:56]: Sure. Nice idea.

File: src/main/java/org/simulator/plot/PlotProcessedSedmlResults.java [hunk 18]
```diff
@@ -37,117 +37,102 @@
...
-        
-        return pwd.getParentFile().getAbsolutePath();
-	}
-}
\ No newline at end of file
+  /**
+   * Get absolute parent path for given test resource.
+   * Due to the relative paths of SBML and SED-ML files the resource loading is not working
+   * in maven.
+   * <p>
```
[the remark this answers is not in the document]
Comment (hemilpanchiwala) [reply src/main/java/org/simulator/plot/PlotProcessedSedmlResults.java:127]: @zakharc, it gives example of how to input the parameter.

File: src/main/java/org/testsuite/SBMLTestSuiteRunnerWrapper.java [hunk 19]
```diff
@@ -0,0 +1,237 @@
...
+                amountHash.put(s, true);
+            }
+        }
+
+        for (String s : concentrations) {
+            s = s.trim();
+            if (!s.isEmpty()) {
+                amountHash.put(s, false);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (hemilpanchiwala) [reply src/main/java/org/testsuite/SBMLTestSuiteRunnerWrapper.java:99]: @zakharc, this class is not for testing but instead is used for getting the results in the SBML Test Runner. But yes, I will make this class more cleaner by diving the code into multiple methods.

File: src/main/java/org/testsuite/SBMLTestSuiteRunnerWrapper.java [hunk 20]
```diff
@@ -0,0 +1,237 @@
...
+        MultiTable inputData = csvimporter.convert(model, resultsPath);
+        double[] timepoints = inputData.getTimePoints();
+        duration = timepoints[timepoints.length - 1]
+                - timepoints[0];
+
+        MultiTable solution;
+
+        // writes results to the output file in CSV format
+        File outputFile = new File(outputFilePath);
+        outputFile.createNewFile();
```
[the remark this answers is not in the document]
Comment (hemilpanchiwala) [reply src/main/java/org/testsuite/SBMLTestSuiteRunnerWrapper.java:117]: @zakharc, No, it doesn't need exception as it just deletes an old file (if it already exists) and add new one to the desired location.

File: src/main/java/org/simulator/plot/PlotProcessedSedmlResults.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (zakharc) [reply src/main/java/org/simulator/plot/PlotProcessedSedmlResults.java:127]: The method signature implies that we want to get a folder path for a test resource and return a string. The javadoc explains that we get an absolute parent path for a given test resource. The "Example: ..."  gives an example of how to input the parameter. But the comment "Due to the relative paths of SBML and SED-ML files the resource loading is not working in maven." has no relation to the method at all.

If we want to explicitly show, that the relative path is awaited as a parameter I would simply explain this in the corresponding @param description of javadoc.

File: src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java (hunk 12 above)
Comment (hemilpanchiwala) [reply src/test/java/org/simulator/sbml/SBMLTestSuiteTest.java:200]: @zakharc, as there is an another import used `import org.sbml.jsbml.xml.stax.SBMLReader` that's why it is using full canonical name.

File: src/main/java/org/testsuite/SBMLTestSuiteRunnerWrapper.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (zakharc) [reply src/main/java/org/testsuite/SBMLTestSuiteRunnerWrapper.java:117]: @hemilpanchiwala: The procedure itself can be not allowed under certain circumstances. Example: 

- We want to delete the file (as it just deletes an old file) and this is not allowed due to the changed writing rights (chmod 555 on linux etc.). 
- The CI system executes your code as a third-person user and has no credentials to do this.

For me, it looks like this should result in an exception. I am not sure what API's we are using "under the hood" (java.io.File, java.nio.file.Paths, ore something else). But to my understanding - every single one of them will result in an exception.

File: src/test/scripts/sbml_test_suite_runner_wrapper.sh (hunk 16 above)
Comment (hemilpanchiwala) [reply src/test/scripts/sbml_test_suite_runner_wrapper.sh]: I just added the needed jars in the sbml_test_runner_wrapper folder which made things cleaner.

**Comment by hemilpanchiwala** [comment]
@zakharc, I have updated most of the requested changes. I will soon update the SBMLTestSuiteRunnerWrapper class by dividing the code into different methods. And for the JavaDoc comments of the overridden methods and some others, I will update it as I work. However, all comment changes pointed by you have been updated.

**Comment by shalinshah1993** [comment]
Lots of good work here! 

- For the white space diffs, it seems that they can be hidden away by clicking on settings and avoid white space changes. However, to avoid pushing them in commits in future, you need to change git attributes. Please add these lines in the `.gitattributes` file. It is the default template for Java code to alway convert the code to and will avoid the white space diffs. You can find default template for other languages here - https://github.com/github/gitignore 

```
# avoid
text=auto 
eol=lf

# Java sources
*.java          text diff=java
*.gradle        text diff=java
*.gradle.kts    text diff=java

# These files are text and should be normalized (Convert crlf => lf)
*.css           text diff=css
*.df            text
*.htm           text diff=html
*.html          text diff=html
*.js            text
*.jsp           text
*.jspf          text
*.jspx          text
*.properties    text
*.tld           text
*.tag           text
*.tagx          text
*.xml           text

# These files are binary and should be left untouched
# (binary is a macro for -text -diff)
*.class         binary
*.dll           binary
*.ear           binary
*.jar           binary
*.so            binary
*.war           binary
*.jks           binary
```

- I saw some coding convention pointed out up here, while I agree it is a good practice to adhere to them, I was thinking may be we can incorporate style checks. For python, I know `pylint` can be easily incorporated with Jenkins. A quick search for java showed checkstyle. https://checkstyle.sourceforge.io/ @draeger how the timeline for this project, may be if it is really easy, we can incorporate lint checker with maven and travis? 

- I also see that that one test is failing. Before this gets merged, can you either fix the test or for now comment it out so it doesn't break the `master` branch?

**Comment by hemilpanchiwala** [comment]
@shalinshah1993, I will add the snippet you gave in the `.gitattributes` file in the next commit. But for the tests, I don't think there is a single failing test instead there are many SBML Test Suite tests that are failing currently due to some SBML L3V2 and other issues.

File: src/main/java/org/testsuite/SBMLTestSuiteRunnerWrapper.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (zakharc) [reply src/main/java/org/testsuite/SBMLTestSuiteRunnerWrapper.java:102]: I was implying, that this piece of code could go to a separate object. Let's say SBMLTestRunnerSuiteConfiguration. In this way, we can encapsulate configuration initialization in a separate class. As a result, we can test this with a separate component test. Let's say SBMLTestRunnerSuiteConfigurationTest. It is an idea for the future refactoring, that you might want to do.

But on this stage, I would suggest dealing with failing tests first. So I will mark this as resolved.

**Comment by hemilpanchiwala** [comment]
> Why did subclassing not work? I.e. only overwriting the `solve` method?
@matthiaskoenig,
It did not work as the new instance of GlpkSolver was made in the solve() method explicitly. We did not have any reference to that class in SBSCL. 

Here, the thing is going on as follows:
We make an object of LinearProgramSolver (which is an interface) that is implemented by GLPKSolver class where the solve() method implementation is found. In this solve() method, scpsolver creates an instance of the GlpkSolver class which was creating all the problems and so making that instance (of GlpkSolver) null using deleteProb() solves the issue. But this is all in scpsolver so I made a copy of the GLPKSolver class with making the instance of GlpkSolver null on completing its usage. Once scpsolver makes this change then we can remove this class.

**Note:** GLPKSolver and GlpkSolver are different classes in scpsolver.

File: src/main/java/org/simulator/examples/SimulatorExample.java [hunk 21]
```diff
@@ -31,78 +31,96 @@
...
-    SBMLDocument document = (new SBMLReader()).readSBML(fileName);
-    Model model = document.getModel();
-    
-    DESSolver solver = new RosenbrockSolver();
-    solver.setStepSize(stepSize);
-    SBMLinterpreter interpreter = new SBMLinterpreter(model);
-    if (solver instanceof AbstractDESSolver) {
-      ((AbstractDESSolver) solver).setIncludeIntermediates(false);
-    }
+        String fileName = "";
```
Comment (matthiaskoenig) [inline src/main/java/org/simulator/examples/SimulatorExample.java:75]: Normally one initializes with `null`. Empty string is always a very risky choice. Change to `null`.

File: src/main/java/org/simulator/fba/FluxBalanceAnalysis.java [hunk 22]
```diff
@@ -50,433 +50,419 @@
...
+    this.eps = eps;
+  }
+
+  /**
+   * Gets the id of the active objective function
+   *
+   * @return
+   */
+  public String getActObjFunc() {
+    return actObjFunc;
```
Comment (matthiaskoenig) [inline src/main/java/org/simulator/fba/FluxBalanceAnalysis.java:466]: Better name is `activeObjective`. Always write words out. There is code completion which makes larger names easy to use. Also rename functions to `getActiveObjective`.

**Comment by matthiaskoenig** [comment]
Will merge this as soon as the minor changes & the test case are there.

File: src/main/java/org/simulator/fba/FluxBalanceAnalysis.java (hunk 22 above)
Comment (hemilpanchiwala) [reply src/main/java/org/simulator/fba/FluxBalanceAnalysis.java]: Updated!

File: src/main/java/org/simulator/examples/SimulatorExample.java (hunk 21 above)
Comment (hemilpanchiwala) [reply src/main/java/org/simulator/examples/SimulatorExample.java]: Updated the code!

**Comment by hemilpanchiwala** [comment]
> Please add a FluxBalanceAnalysis test which:
> 
> * takes the ecoli core model, simulates a FBA, compares objective; currently a simple test case for FBA outside of the SBMLTestSuite is missing

@matthiaskoenig, I have updated the changes. And FBA test was there but it was named CobraSolverTest and was ignored by @Ignore tag. I have renamed it to FluxBalanceAnalysisTest and also removed the ignore tag.

**Pull request merged**